// marmiton-api/dist/lib/components/recipes-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RecipesParser = void 0;
const node_html_parser_1 = require("node-html-parser");
const recipe_builder_1 = require("./recipe-builder");
const node_fetch_1 = require("node-fetch");
const recipe_enums_1 = require("./recipe-enums");
class RecipesParser {
    static async parseSearchResults(dom, baseUrl) {
        const recipes = await Promise.all((0, node_html_parser_1.parse)(dom)
            .querySelectorAll('main div > a')
            .filter((e) => e.getAttribute('class')?.includes('MRTN'))
            .map(async (e) => {
            const url = new URL(e.getAttribute('href').trim(), baseUrl);
            // Get as many info as we can from the recipe search result
            // These cannot be undefined.
            const rb = new recipe_builder_1.RecipeBuilder()
                .withName(this.selectText(e, 'h4'))
                // .withDescription(this.selectText(e, '.recipe-card__description'))
                .withRate(Number(this.selectText(e, 'span')?.replace(/\/\s*5/, '')))
                .withUrl(url.toString());
            // Load the recipe page
            const dom = await (await (0, node_fetch_1.default)(url.toString())).text();
            try {
                return await this.parseRecipe(dom, rb);
            }
            catch {
                // Some recipe are ads ?
                return undefined;
            }
        }));
        return recipes.filter((r) => r !== undefined);
    }
    static async parseRecipe(dom, rb = new recipe_builder_1.RecipeBuilder()) {
        // Set "raw" attributes. These doesn't need any processing steps
        const recipeRoot = (0, node_html_parser_1.parse)(dom);
        let recipe = recipeRoot
            .querySelectorAll('script[type="application/ld+json"]')
            .map((el) => JSON.parse(this.getCleanText(el.childNodes[0])))
            // Not really needed but just to be sure
            .find((r) => r['@type'] === 'Recipe');
        if (recipe === undefined)
            return;
        // Gather every raw attributes we can
        rb.withIngredients(recipe?.recipeIngredient)
            .withAuthor(recipe?.author)
            // This attribute can either be a string or a string array, normalize it
            .withImages([recipe?.image].flat())
            .withSteps(recipe?.recipeInstructions.map((ri) => ri?.text))
            .withDescription(recipe?.description);
        // French attributes.
        const keywordsArray = recipe?.keywords.split(/,\s*/);
        // This is a very shady way to retrieve these two data.
        // However this is the only related data in the object ?
        const rawBudget = keywordsArray[keywordsArray.length - 1];
        const rawDifficulty = keywordsArray[keywordsArray.length - 2];
        rb.withDifficulty(this.parseDifficulty(rawDifficulty))
            .withBudget(this.parseBudget(rawBudget))
            .withTags(keywordsArray);
        // Time related attributes
        rb.withPreparationTime(this.parseISO8601(recipe?.prepTime)).withTotalTime(this.parseISO8601(recipe?.totalTime));
        // "Optional" attributes
        // Pure regex parsing isn't that consistent, better prepare for the worst
        const people = Number(recipe.recipeYield.match(/\d+/)[0]);
        rb.withPeople(people);
        return rb.build();
    }
    /**
     * Parse an ISO 8601 string and return a duration in minutes.
     * @param duration
     * @private
     */
    static parseISO8601(duration) {
        // I really don't want to include moment just for this
        const matches = duration.match(this.ISO_8601_REGEX);
        // We're going to assume that the granularity is in minutes
        // And that a recipe don't take more than a few hours ?
        const minutes = matches?.[matches.length - 2]?.match(/\d+/)?.[0] ?? 0;
        const hours = matches?.[matches.length - 3]?.match(/\d+/)?.[0] ?? 0;
        return Number(hours) * 60 + Number(minutes);
    }
    /**
     * Converts french textual representation of a recipe budget to an enum
     * @param budget
     */
    static parseBudget(budget) {
        switch (budget.toLowerCase()) {
            case 'bon marché':
                return recipe_enums_1.RECIPE_PRICE.CHEAP;
            case 'moyen':
                return recipe_enums_1.RECIPE_PRICE.MEDIUM;
            case 'assez cher':
                return recipe_enums_1.RECIPE_PRICE.EXPENSIVE;
            default:
                return recipe_enums_1.RECIPE_PRICE.MEDIUM;
        }
    }
    /**
     * Converts french textual representation of a recipe difficulty to an enum
     * @param budget
     */
    static parseDifficulty(difficulty) {
        switch (difficulty.toLowerCase()) {
            case 'très facile':
                return recipe_enums_1.RECIPE_DIFFICULTY.VERY_EASY;
            case 'facile':
                return recipe_enums_1.RECIPE_DIFFICULTY.EASY;
            case 'moyenne':
                return recipe_enums_1.RECIPE_DIFFICULTY.MEDIUM;
            case 'difficile':
                return recipe_enums_1.RECIPE_DIFFICULTY.HARD;
            default:
                return recipe_enums_1.RECIPE_DIFFICULTY.MEDIUM;
        }
    }
    static selectText(root, selector) {
        return this.getCleanText(root.querySelector(selector));
    }
    static getCleanText(e) {
        return (e?.text
            ?.trim()
            // Replace unicode codes with characters
            .replace(/\\u([\d\w]{4})/gi, (match, grp) => String.fromCharCode(parseInt(grp, 16)))
            // Remove all tabs, and linebreaks
            .replace(/[\n\r\t]/g, '')
            // Allow only a single space between words
            .replace(/\s+/, ' ')
            // If there is no space between the first digit and a word (ingredients)
            // add it
            .replace(/(\d)([A-Za-z]{2,})/, '$1 $2'));
    }
}
exports.RecipesParser = RecipesParser;
/** ISO 8601 Regex. The only capture groups used for a recipe should be H and M */
RecipesParser.ISO_8601_REGEX = /^P(?!$)(\d+(?:\.\d+)?Y)?(\d+(?:\.\d+)?M)?(\d+(?:\.\d+)?W)?(\d+(?:\.\d+)?D)?(T(?=\d)(\d+(?:\.\d+)?H)?(\d+(?:\.\d+)?M)?(\d+(?:\.\d+)?S)?)?$/;
